src/rust/collections/concurrent_ring.rs

```diff
@@ -0,0 +1,509 @@
+    /// the amount of storage space in bytes that should be available. [size] must be at least large enough to hold
+    /// capacity, plus the push and pop offsets and a [HEADER_SIZE] piece of padding.
+    pub fn from_raw_parts(init: bool, mut ptr: *mut u8, capacity: usize) -> Result<Self, Fail> {
+        // Check if we have a valid pointer.
+        if ptr.is_null() {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "cannot construct a ring buffer from a null pointer",
+            ));
+        }
+
+        // Check if the memory region is aligned to usize.
+        let align_of_usize: usize = mem::align_of::<usize>();
+        if ptr.align_offset(align_of_usize) != 0 {
+            return Err(Fail::new(
+                libc::EINVAL,
+                "cannot construct a ring buffer from a unaligned memory region",
+            ));
+        }
+
+        const SIZE_OF_USIZE: usize = mem::size_of::<usize>();
+        // Check that there is sufficient space in the buffer.
+        let size_of_ring: usize = 2 * SIZE_OF_USIZE;
+
+        // Compute pointers and required padding.
+        let pop_offset: *mut usize = ptr as *mut usize;
+        unsafe { ptr = ptr.add(SIZE_OF_USIZE) };
+        let push_offset: *mut usize = ptr as *mut usize;
+        unsafe { ptr = ptr.add(SIZE_OF_USIZE) };
+        let buffer_ptr: *mut u8 = ptr as *mut u8;
+
+        // Initialize enqueue and dequeue pointers only if requested.
+        if init {
+            unsafe {
+                *push_offset = 0;
+                *pop_offset = 0;
+            }
+        }
+
+        Ok(Self {
+            push_offset,
+            pop_offset,
+            buffer: raw_array::RawArray::<u8>::from_raw_parts(buffer_ptr, capacity - size_of_ring)?,
+            is_managed: false,
+        })
+    }
+
+    /// Returns the effective capacity of the target ring buffer.
+    #[allow(unused)]
+    pub fn capacity(&self) -> usize {
+        self.buffer.capacity()
+    }
+
+    /// Attempts to insert a buffer of [len] bytes into the ring buffer.
+    pub fn try_push(&self, buf: &[u8]) -> Result<(), Fail> {
+        let len: usize = buf.len();
+        if len == 0 {
+            return Err(Fail::new(libc::EINVAL, "Buffer must be non-zero length"));
+        }
+        if let Some(push_offset) = self.allocate_space(len + HEADER_SIZE) {
+            let buf_ptr: *const u8 = buf.as_ptr();
+            unsafe {
+                // Pointer to where to start writing data for this message.
+                let push_ptr: *mut u8 = self.buffer.get_mut().as_mut_ptr().add(push_offset + HEADER_SIZE);
+                // Copy the data into the ring buffer.
+                copy(buf_ptr, push_ptr, len);
+            }
+            // Commit the write by atomically writing the header to release the buffer. The overwritten header MUST be 0.
+            assert_eq!(self.write_header(len, push_offset), 0);
```

Comment:
Should this be turned into a `debug_assert!()` since it is not allowed by construction?